### PR TITLE
Bug 1222131 - Adding Gecko revision to /device

### PR DESCRIPTION
**Pull request by eliperelman**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "lodash": "^3.10.1",
     "moment": "^2.10.6",
     "mz": "^2.0.0",
+    "sax": "^1.1.4",
```
Comment (gaye) [inline package.json]: uh oh here we go

Comment (eliperelman) [inline package.json]: LOOOL :laughing:

File: src/fxos/readGeckoCommit.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+let fs = require('mz/fs');
+let path = require('path');
+let sax = require('sax');
+let tmpdir = denodeify(require('tmp').dir);
+
+const SOURCES_XML = '/system/sources.xml';
+const APPLICATION_INI = '/system/b2g/application.ini';
+const PLATFORM_INI = '/system/b2g/platform.ini';
+
+function pullIniFile(adb, local, file) {
```
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: async/await man!

Comment (eliperelman) [inline src/fxos/readGeckoCommit.js]: You'll see why I went this route, the code for async/await in this case I think makes it look a lot more complex. Let me know if you still want it changed.

File: src/fxos/readGeckoCommit.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    stream.on('data', buffer => {
+      let line = buffer
+        .split('\n')
+        .find(line => line.includes('SourceStamp'));
+
+      if (!line) {
+        return;
+      }
+
+      revision = line.split('=')[1];
```
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: Don't you want to remove the rejection listener here or check the revision against a regex or something?

Comment (gaye) [inline src/fxos/readGeckoCommit.js]: Also when you resolve you should remove the `'data'` listener

Comment (eliperelman) [inline src/fxos/readGeckoCommit.js]: Both of those suggestions make sense, I'll make the change.

File: src/fxos/readGeckoCommit.js (hunk 2 above)
Comment (eliperelman) [inline src/fxos/readGeckoCommit.js]: Oh nevermind, I thought you were talking about a different section. Gah, I'll change!

File: src/fxos/readGeckoCommit.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        reject('Unable to find Gecko revision in INI file');
+      }
+    });
+  });
+}
+
+function getGeckoRevisionFromIni(adb, local) {
+  return Promise
+    .resolve()
+    .then(() => pullIniFile(adb, local, APPLICATION_INI))
```
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: moar async

Comment (gaye) [inline src/fxos/readGeckoCommit.js]: Why don't you try both paths concurrently like we did for crashes?

File: src/fxos/readGeckoCommit.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+function getGeckoRevisionFromXml(local) {
+  let revision;
+
+  return new Promise((resolve, reject) => {
+    let stream = sax.createStream();
+
+    // The Gecko revision lives in a tag that MAY be HTML-commented
+    stream.on('opentag', node => {
```
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: So this is the same general pattern where you have

```
1. A stream
2. Something that you expect to receive in the stream
3. You want to fail if the stream ends and you don't see the thing
```

Could we do something like `await findInStream(stream, predicate);`?

Comment (gaye) [inline src/fxos/readGeckoCommit.js]: Basically what I am saying is that this is a lot of code for what should be something like the `Stream.find` to `Array.find`

Comment (eliperelman) [inline src/fxos/readGeckoCommit.js]: I'd agree with you on text streams where the data comes from the `data` event, but with object-based streams like this one, the data comes from the `opentag` event and I don't see any real easy way to abstract it the same way. I'll chew on it to see if I can think of anything.

Comment (eliperelman) [inline src/fxos/readGeckoCommit.js]: I guess we can pass another parameter to override the event type.

Comment (gaye) [inline src/fxos/readGeckoCommit.js]: ++ That

File: src/find-in-stream.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    stream.on('end', end);
+    stream.on(eventType, function handler(item) {
+      if (predicate(item)) {
+        stream.removeListener(eventType, handler);
+        stream.removeListener('end', end);
+        return resolve(item);
+      }
+    });
+  });
+};
```
Comment (gaye) [inline src/find-in-stream.js]: Perfect!

Comment (gaye) [inline src/find-in-stream.js]: Maybe `findInStream.js` so our naming is consistent?

File: src/fxos/readGeckoCommit.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+function matchSourceStamp(item) {
+  return item
+    .split('\n')
+    .find(line => line.includes('SourceStamp'));
+}
+
+async function readGeckoRevisionFromIni(iniFile) {
+  let stream = fs.createReadStream(iniFile, {encoding: 'utf8'});
+  let line = await findInStream(stream, 'data', matchSourceStamp);
```
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: So much cleaner! What do you think about `'data'` a default value?

Comment (eliperelman) [inline src/fxos/readGeckoCommit.js]: If we did that we would need to reorder the args. If we did that I'd want predicate first.

File: src/fxos/readGeckoCommit.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  return matchSourceStamp(line).split('=')[1];
+}
+
+async function getGeckoRevisionFromIni(adb, local) {
+  let iniFile;
+
+  await Promise.all([APPLICATION_INI, PLATFORM_INI].map(async function(file) {
+    try {
+      iniFile = await pullIniFile(adb, local, file);
```
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: This looks dangerous!

File: src/fxos/readGeckoCommit.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  return await readGeckoRevisionFromIni(iniFile);
+}
+
+async function getGeckoRevisionFromXml(local) {
+  let stream = fs.createReadStream(local).pipe(sax.createStream());
+  let node = await findInStream(stream, 'opentag', node => {
+    // The Gecko revision lives in a tag that MAY be HTML-commented
+    return node.name === 'PROJECT' && node.attributes.NAME === 'gecko';
```
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: ++

File: src/device.js [hunk 10]
```diff
@@ -7,10 +7,11 @@ module.exports = async function device(req, res) {
...
 
-  let [gaiaCommit] = await Promise.all([
-    fxos.readGaiaCommit(adb)
+  let [gaiaCommit, geckoCommit] = await Promise.all([
+    fxos.readGaiaCommit(adb),
+    fxos.readGeckoCommit(adb)
   ]);
 
-  let result = Object.assign({}, {gaia: gaiaCommit});
+  let result = Object.assign({}, {gaia: gaiaCommit, gecko: geckoCommit});
```
Comment (gaye) [inline src/device.js]: Do we still need this `Object.assign` call?

File: src/fxos/readGeckoCommit.js (hunk 7 above)
Comment (gaye) [inline src/fxos/readGeckoCommit.js]: How about `findInStream(stream, matchSourceStamp, {eventType: 'openTag})` and default last arg is `{eventType: 'data'}`?

Comment (eliperelman) [inline src/fxos/readGeckoCommit.js]: Sure!

**Pull request merged**